Check if attachment is actually(!) referred to

pabzm opened the pull request:
This finishes what #9472 intended to do – but didn't actually do, as I found out.

The code now checks for each non-text mime-part in a multipart-part if its `Content-ID` or `Content-Location` is (probably) used in a sibling HTML-part, and only if that matches the respective mime-part is considered an "inline" attachment (that won't show up as downloadable or below the message content).

The second commit in this PR makes sure that for all image-parts, all mime-part-headers are loaded from the server, in order to actually get hands on the `Content-Location`-header (which isn't always fetched in the first place). It is limited to image-parts because those are the most common ones maybe having a `Content-Location`-header and I assumed that we shouldn't load the headers for every mime-part, so this seems like a workable real-world distinction for me.

One could probably change how the `BODYSTRUCTURE` response is fetched and parsed to ensure a `Content-Location`-header is always fetched in the first place, but I didn't dare to touch that code.

This fix closes #9565

pabzm committed:
Check if "inline" msg part is actually referred to

If there's no reference to it in a sibling HTML part then we handle it
as a classic attachment (which is shown as downloadable).

pabzm committed:
Fetch all msg headers also for images to always get Content-Location

Previously all headers were only fetched for message/rfc822, or
if the Content-Type's "name" parameter was set, or if a Content-ID was
set.
The RFC doesn't require neither the "name" parameter nor a Content-ID
for using Content-Location, though, so we shouldn't depend on those.

Instead now all headers are also fetched if the main part of the
Content-Type is "image", to catch more cases.

alecpl commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -1006,13 +1006,37 @@ private function parse_structure($structure, $recursive = false)
+                        }
+                        // Note: There might be more than one HTML part, thus
+                        // we use a callback and concatenate the results.
+                        $html_content = implode('', array_map(function ($part) { return $this->get_part_body($part->mime_id); }, $html_parts));
```
There's indeed my @TODO comment 70 lines below to get the HTML bodies and check for references. However, this has a performance impact, that for a message with many images and big HTML content might be noticeable. As of now I considered fetching part headers acceptable, but part body is another story.

We need some more considerations. For example, when loading an image attachment (`rcmail_attachment_handler`) maybe we could fetch the image without needing to parse the message structure and loading HTML parts again only to get the attachment part data.

Maybe checking for references must be done outside of `rcube_message`. So `rcube_message` is not that heavy (to not slow down all the cases where we deal with the message), but do not really need the full structure information, e.g. when viewing source or downloading the message, or when dealing with a single attachment.

Or maybe we need to use cache. It might not help much when dealing with parallel requests (loading image attachments) though. And caches are usually optional.

pabzm replied:
I see your point and am working on it.

olegStreejak replied:
FYI https://github.com/roundcube/roundcubemail/issues/9565#issuecomment-2327319524

pabzm replied:
I changed the approach so the reference checking will be only done on demand. I'd like to test this, but I have a hard time figuring out how, since the entire class essentially depends on the response to an IMAP `bodystructure` command, which I'd like to avoid mocking.

pabzm replied:
I made some progress and am using the approach in #9460 for testing, but this needs more tests to be ready. I'm mostly afk until the beginning of January, though.

pabzm committed:
Parse HTML for references only on demand

pabzm committed:
Typos and comment formatting

pabzm committed:
Don't skip test anymore

We want it tested!

pabzm committed:
More MR tests with images

pabzm committed:
Remove early special handling for "inline" images

We decide later, which attachment is considered "inline" and which
isn't.

pabzm committed:
Remove early resolving of references in TNEF parts

pabzm committed:
Testing message rendering of TNEF emails

pabzm committed:
Don't use image disposition, it's unreliable

pabzm committed:
Split adding raw parts and attachments

pabzm committed:
Fix renaming variable

pabzm committed:
Rename file to make its test be run

pabzm committed:
Remove outdated script

pabzm committed:
Annotate test cases with GitHub issue numbers

pabzm committed:
Fix test case class name

pabzm committed:
remove comment

pabzm committed:
Test inline image message rendering

pabzm committed:
Rename test file to reflect cases better

pabzm commented:
I would like to integrate a lot more messages into the "Message Rendering" tests, but the rest of the code is ready as it is, and I can supplement additional tests later, too.

alecpl commented on tests/MessageRendering/data/greenmail/test-message-rendering@localhost/INBOX/spam-with-image-data-uri.eml:
```diff
@@ -0,0 +1,6734 @@
+Content-ID: <2478d4f4b6b373ac9292cdbab380ee6bd4e4a16ee4feef5112ac3fe61b1303fe>
+Content-Transfer-Encoding: base64
+
+iVBORw0KGgoAAAANSUhEUgAAAlsAAAKYCAYAAABEuT1rAAAAAXNSR0IArs4c6QAAAARnQU1BAACx
```
It would be good to keep the test data small. I.e. use some simple/small images.

pabzm replied:
I was using original emails, but I see the point, we can save quite some byte here 👍

alecpl reviewed (commented):
I'll need more time to properly test this.

alecpl commented:
I've tested a bit with various messages. I've found one regression: a forwarded message with eml attachment - there's a difference in displaying it with this patch and without. The attached message headers section is missing. Also if you open the eml attachment the headers will be missing.

ps. I noticed that we FETCH BODY.PEEK[2] for such messages, i.e. we fetch the whole attached message. It seems not optimal if it contains big attachments (I'd have to investigate that separately) - it's the same in release-1.6, so nothing new, I suppose.

pabzm committed:
Reduce image used in test email

It doesn't change much, but there's also no sense in decoding big images
that we don't use.

pabzm committed:
Remove unused variable initialisation

pabzm commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -1111,27 +1150,28 @@ private function get_mime_numbers(&$part)
             // use part ID index to prevent from duplicates
-            switch ($type) {
-                case 'inline':
-                    $this->inline_parts[$part->mime_id] = $part;
```
Good catch, thank you!

pabzm commented:
> I've found one regression: a forwarded message with eml attachment

Thank you for finding this! It's present in the "master" branch already, though. I'll try to find out which change is to blame and how to fix it.

alecpl commented:
> > I've found one regression: a forwarded message with eml attachment
> 
> Thank you for finding this! It's present in the "master" branch already, though. I'll try to find out which change is to blame and how to fix it.

It works fine in master.

pabzm commented:
> It works fine in master.

I double-checked and still see the wrong behaviour on the lastest state of the "master"-branch.

I'm testing with the attached message, can you please check if it's displayed properly in your setup?

[Fwd Lines.eml.zip](https://github.com/user-attachments/files/18424684/Fwd.Lines.eml.zip)

And can you provide me with your test message?

alecpl commented:
Ah, sorry. You're right. It works in release-1.6, but not in master. My bad.

pabzm commented:
This is what I see with Roundcubemail v1.6.9:

<kbd>
<img width="736" alt="Screenshot RC 1.6.9" src="https://github.com/user-attachments/assets/539f3552-846f-4ade-8a09-deb64825ae36" />
</kbd>

---------

And this is what I see with Roundcubemail "master" branch:

<kbd>
<img width="784" alt="Screenshot RC 'master' branch" src="https://github.com/user-attachments/assets/6c1d7913-7787-40be-bc10-591af94cb7cb" />
</kbd>

pabzm commented:
Actually my fix does fix the bug if I work thoroughly: https://github.com/roundcube/roundcubemail/pull/9753

pabzm commented:
Rebased onto the "master" branch to fix a conflict
